File: mobile-rn/src/components/AudioRecorderField.js
import { useState } from "react";
import { Pressable, StyleSheet, Text, View } from "react-native";
import {
  RecordingPresets,
  requestRecordingPermissionsAsync,
  setAudioModeAsync,
  useAudioPlayer,
  useAudioPlayerStatus,
  useAudioRecorder,
  useAudioRecorderState
} from "expo-audio";
import { colors, radius } from "../theme";

// uri: string | null — the recorded voice-note file, or null if none yet.
// onChange(uri | null) — called once recording stops, or when the user removes it.
export default function AudioRecorderField({ uri, onChange }) {
  const recorder = useAudioRecorder(RecordingPresets.HIGH_QUALITY);
  const recorderState = useAudioRecorderState(recorder);
  const player = useAudioPlayer(uri || undefined);
  const playerStatus = useAudioPlayerStatus(player);
  const [error, setError] = useState("");

  async function startRecording() {
    setError("");
    const { granted } = await requestRecordingPermissionsAsync();
    if (!granted) {
      setError("Autorisez l'accès au micro pour enregistrer une note vocale.");
      return;
    }
    await setAudioModeAsync({ allowsRecording: true, playsInSilentMode: true });
    await recorder.prepareToRecordAsync();
    recorder.record();
  }

  async function stopRecording() {
    await recorder.stop();
    if (recorder.uri) onChange(recorder.uri);
  }

  function togglePlayback() {
    if (playerStatus.playing) player.pause();
    else player.play();
  }

  const seconds = Math.floor((recorderState.durationMillis || 0) / 1000);

  return (
    <View style={styles.wrap}>
      {uri ? (
        <View style={styles.previewRow}>
          <Pressable style={styles.playBtn} onPress={togglePlayback}>
            <Text style={styles.playBtnText}>{playerStatus.playing ? "⏸" : "▶"}</Text>
          </Pressable>
          <Text style={styles.previewLabel}>Note vocale enregistrée</Text>
          <Pressable style={styles.removeBtn} onPress={() => onChange(null)}>
            <Text style={styles.removeBtnText}>✕</Text>
          </Pressable>
        </View>
      ) : recorderState.isRecording ? (
        <Pressable style={[styles.recordBtn, styles.recording]} onPress={stopRecording}>
          <View style={styles.recDot} />
          <Text style={[styles.recordBtnText, styles.recordingText]}>Arrêter l'enregistrement · {seconds}s</Text>
        </Pressable>
      ) : (
        <Pressable style={styles.recordBtn} onPress={startRecording}>
          <View style={styles.camDot} />
          <Text style={styles.recordBtnText}>🎤 Ajouter une note vocale (facultatif)</Text>
        </Pressable>
      )}
      {!!error && <Text style={styles.error}>{error}</Text>}
    </View>
  );
}

const styles = StyleSheet.create({
  wrap: { marginBottom: 16 },
  recordBtn: {
    flexDirection: "row",
    alignItems: "center",
    justifyContent: "center",
    gap: 8,
    height: 42,
    borderWidth: 1.5,
    borderColor: colors.hairline,
    borderRadius: radius
  },
  recording: { borderColor: colors.error },
  recordBtnText: { color: colors.teal, fontSize: 12 },
  recordingText: { color: colors.error },
  camDot: { width: 7, height: 7, borderRadius: 4, backgroundColor: colors.teal },
  recDot: { width: 7, height: 7, borderRadius: 4, backgroundColor: colors.error },
  previewRow: {
    flexDirection: "row",
    alignItems: "center",
    gap: 10,
    height: 42,
    paddingHorizontal: 12,
    borderWidth: 1.5,
    borderColor: colors.hairline,
    borderRadius: radius
  },
  playBtn: {
    width: 26,
    height: 26,
    borderRadius: 13,
    backgroundColor: colors.amber,
    alignItems: "center",
    justifyContent: "center"
  },
  playBtnText: { color: colors.amberOn, fontSize: 11 },
  previewLabel: { flex: 1, color: colors.textDim, fontSize: 12 },
  removeBtn: { width: 24, height: 24, alignItems: "center", justifyContent: "center" },
  removeBtnText: { color: colors.textDim, fontSize: 12 },
  error: { color: colors.error, fontSize: 11, marginTop: 6 }
});

File: mobile-flutter/lib/widgets/audio_recorder_field.dart
import 'dart:async';
import 'dart:io';

import 'package:audioplayers/audioplayers.dart' as ap;
import 'package:flutter/material.dart';
import 'package:path_provider/path_provider.dart';
import 'package:record/record.dart';

import '../theme.dart';

// file: the recorded voice-note, or null if none yet.
// onChanged(file | null): called once recording stops, or when the user removes it.
class AudioRecorderField extends StatefulWidget {
  final File? file;
  final ValueChanged<File?> onChanged;

  const AudioRecorderField({super.key, required this.file, required this.onChanged});

  @override
  State<AudioRecorderField> createState() => _AudioRecorderFieldState();
}

class _AudioRecorderFieldState extends State<AudioRecorderField> {
  final _recorder = AudioRecorder();
  final _player = ap.AudioPlayer();
  StreamSubscription<ap.PlayerState>? _playerSub;
  Timer? _timer;

  bool _recording = false;
  bool _playing = false;
  Duration _elapsed = Duration.zero;
  String _error = '';

  @override
  void initState() {
    super.initState();
    _playerSub = _player.onPlayerStateChanged.listen((state) {
      if (mounted) setState(() => _playing = state == ap.PlayerState.playing);
    });
  }

  @override
  void dispose() {
    _timer?.cancel();
    _playerSub?.cancel();
    _recorder.dispose();
    _player.dispose();
    super.dispose();
  }

  Future<void> _start() async {
    setState(() => _error = '');
    if (!await _recorder.hasPermission()) {
      setState(() => _error = "Autorisez l'accès au micro pour enregistrer une note vocale.");
      return;
    }
    final dir = await getTemporaryDirectory();
    final path = '${dir.path}/voice-${DateTime.now().millisecondsSinceEpoch}.m4a';
    await _recorder.start(const RecordConfig(), path: path);

    final startedAt = DateTime.now();
    setState(() {
      _recording = true;
      _elapsed = Duration.zero;
    });
    _timer = Timer.periodic(const Duration(milliseconds: 200), (_) {
      if (mounted) setState(() => _elapsed = DateTime.now().difference(startedAt));
    });
  }

  Future<void> _stop() async {
    final path = await _recorder.stop();
    _timer?.cancel();
    if (mounted) setState(() => _recording = false);
    if (path != null) widget.onChanged(File(path));
  }

  Future<void> _togglePlayback() async {
    if (widget.file == null) return;
    if (_playing) {
      await _player.pause();
    } else {
      await _player.play(ap.DeviceFileSource(widget.file!.path));
    }
  }

  Future<void> _remove() async {
    await _player.stop();
    widget.onChanged(null);
  }

  @override
  Widget build(BuildContext context) {
    if (widget.file != null) {
      return Container(
        height: 42,
        padding: const EdgeInsets.symmetric(horizontal: 12),
        decoration: BoxDecoration(borderRadius: BorderRadius.circular(kRadius), border: Border.all(color: AppColors.hairline)),
        child: Row(
          children: [
            InkWell(
              onTap: _togglePlayback,
              borderRadius: BorderRadius.circular(13),
              child: Container(
                width: 26,
                height: 26,
                alignment: Alignment.center,
                decoration: const BoxDecoration(color: AppColors.amber, shape: BoxShape.circle),
                child: Text(_playing ? '⏸' : '▶', style: const TextStyle(color: AppColors.amberOn, fontSize: 11)),
              ),
            ),
            const SizedBox(width: 10),
            const Expanded(child: Text('Note vocale enregistrée', style: TextStyle(color: AppColors.textDim, fontSize: 12))),
            InkWell(
              onTap: _remove,
              child: const Padding(
                padding: EdgeInsets.all(4),
                child: Text('✕', style: TextStyle(color: AppColors.textDim, fontSize: 12)),
              ),
            ),
          ],
        ),
      );
    }

    if (_recording) {
      return InkWell(
        onTap: _stop,
        borderRadius: BorderRadius.circular(kRadius),
        child: Container(
          height: 42,
          alignment: Alignment.center,
          decoration: BoxDecoration(borderRadius: BorderRadius.circular(kRadius), border: Border.all(color: AppColors.error)),
          child: Text("Arrêter l'enregistrement · ${_elapsed.inSeconds}s", style: const TextStyle(color: AppColors.error, fontSize: 12)),
        ),
      );
    }

    return Column(
      crossAxisAlignment: CrossAxisAlignment.stretch,
      children: [
        InkWell(
          onTap: _start,
          borderRadius: BorderRadius.circular(kRadius),
          child: Container(
            height: 42,
            alignment: Alignment.center,
            decoration: BoxDecoration(borderRadius: BorderRadius.circular(kRadius), border: Border.all(color: AppColors.hairline)),
            child: const Text('🎤 Ajouter une note vocale (facultatif)', style: TextStyle(color: AppColors.teal, fontSize: 12)),
          ),
        ),
        if (_error.isNotEmpty)
          Padding(
            padding: const EdgeInsets.only(top: 6),
            child: Text(_error, style: const TextStyle(color: AppColors.error, fontSize: 11)),
          ),
      ],
    );
  }
}
